File: 113.路径总和-ii.cpp
/*
 * @lc app=leetcode.cn id=113 lang=cpp
 *
 * [113] 路径总和 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
#include "algorithm"
#include "iostream"
#include "map"
#include "queue"
#include "vector"

using namespace std;

class Solution {
  private:
    vector<vector<int>> result;

  public:
    vector<vector<int>> pathSum(TreeNode *root, int targetSum) {
        vector<int> v;
        if (!root)
            return {};
        pathSum(root, targetSum, v);
        return result;
    }

    bool isLeaf(TreeNode *root) { return !root->left && !root->right; }

    bool pathSum(TreeNode *root, int targetSum, vector<int> &v) {
        if (isLeaf(root) && targetSum == root->val) {
            v.push_back(root->val);
            result.push_back(v);
            return true;
        }
        v.push_back(root->val);
        targetSum -= root->val;
        if (root->left && pathSum(root->left, targetSum, v)) {
            v.pop_back();
        }
        if (root->right && pathSum(root->right, targetSum, v)) {
            v.pop_back();
        }
        v.pop_back();
        return false;
    }
};
// @lc code=end

File: 113.路径总和-ii.go
/*
 * @lc app=leetcode.cn id=113 lang=golang
 *
 * [113] 路径总和 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) [][]int {
	if root == nil {
		return nil
	}
	var ret [][]int
	var path []int

	var fn func(root *TreeNode, targetSum int)
	fn = func(root *TreeNode, targetSum int) {
		if root == nil {
			return
		}
		defer func() { path = path[:len(path)-1] }()
		path = append(path, root.Val)
		if root.Left == nil && root.Right == nil && targetSum == root.Val {
			ret = append(ret, append([]int(nil), path...))
		}
		fn(root.Left, targetSum-root.Val)
		fn(root.Right, targetSum-root.Val)
	}
	fn(root, targetSum)
	return ret
}

// @lc code=end
